enhancements/monitoring/restart-metrics.md

```diff
@@ -0,0 +1,105 @@
+---
+title: restart-metrics
+authors:
+  - "@rphillips"
+reviewers:
+  - "@smarterclayton"
+  - "@sjenning"
+  - "@mrunalp"
+approvers:
+  - TBD
+creation-date: 2020-03-23
+last-updated: 2020-03-23
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Service Restart Metrics
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+There is not any insight with systemd units restarting. Metrics of this sort are
+important to know if Kubelet, crio, or other system services are crashing and
+restarting. Restart metrics are going to be extremely important since the
+Kubelet is going to be changed to exit on a crash, thus restarting. Previous
+behavior of the Kubelet is to recover() from the crash and not exit.
+
+## Motivation
+
+Systemd unit restart metrics of a unit are vitally important for system
+administrators to see the health of the Kubelet, crio, and other vitally important
+services.
+
+### Goals
+
+- Systemd unit restart metrics need to be propogated through to monitoring, and
+  the alerting system to alert system administrators to problems within the
+  system.
+
+### Non-Goals
+
+### Current Behavior
+
+node_exporter is running in a non-privileged container - as user `nobody` - and
+cannot connect to the DBUS socket to communicate with systemd. Monitoring has
+concerns the underlying [systemd
+collector](https://github.com/prometheus/node_exporter/blob/master/collector/systemd_linux.go)
+is not performant enough.
+
+## Proposal
+
+- This proposal is to raise the privileges on node_exporter.
+
+- Whitelist in the systemd node_exporter collector: crio, kubelet, and any other
+  agreed to system service.
+
+### Risks and Mitigations
+
+We could have a sidecar container running privileged that injects metrics into
+node_exporter; however, increasing privileges on node_exporter is a valid
+option. A new sidecar container would add another dependency and service that
+would need to be installed and maintained.
+
+The systemd collector in node_exporter allows for whitelisted units. To address
+the performance concern, the only services that would be whitelisted would be
+kubelet, crio, and any other core system service that we want systemd metrics
+from.
+
+If there are performance issues with the implementation of this proposal, then
```

Comment:
Have you looked at the other systemd collector? Is there any downsides to that one?

https://github.com/povilasv/systemd_exporter

Reply:
It looks like systemd_exporter would work. I'm curious as to why systemd_exporter is not in node_exporter?

Reply:
>  node-exporter has a broad focus, and process-exporter is focused on deep analysis of specific processes. systemd-exporter aims to fit in the middle, taking advantage of systemd's builtin process and thread grouping to provide application-level metrics.

Guessing they want to keep it separate.

Reply:
I think reasoning upstream is that handling all of systemd is too big/complicated for a single node_exporter collector.  So node_exporter tries to just provide some common default metrics, but if you need more detailed metrics in certain areas the systemd_exporter should be used instead.  Also, communicating with systemd over DBUS can be very slow and resource intensive compared to the other collectors which gather metrics primarily from the procfs filesystem and a few system calls.